internal/app/go-filecoin/storage_market_connector/client.go

```diff
@@ -0,0 +1,229 @@
+func (s *StorageClientNodeConnector) AddFunds(ctx context.Context, addr address.Address, amount tokenamount.TokenAmount) error {
+	return s.addFunds(ctx, s.clientAddr, addr, amount)
+}
+
+// EnsureFunds checks the current balance for an address and adds funds if the balance is below the given amount
+func (s *StorageClientNodeConnector) EnsureFunds(ctx context.Context, addr address.Address, amount tokenamount.TokenAmount) error {
+	balance, err := s.GetBalance(ctx, addr)
+	if err != nil {
+		return err
+	}
+
+	if !balance.Available.LessThan(amount) {
+		return nil
+	}
+
+	return s.AddFunds(ctx, addr, tokenamount.Sub(amount, balance.Available))
+}
+
+// ListClientDeals returns all deals published on chain for the given account
+func (s *StorageClientNodeConnector) ListClientDeals(ctx context.Context, addr address.Address) ([]storagemarket.StorageDeal, error) {
+	return s.listDeals(ctx, addr)
+}
+
+// ListStorageProviders finds all miners that will provide storage
+func (s *StorageClientNodeConnector) ListStorageProviders(ctx context.Context) ([]*storagemarket.StorageProviderInfo, error) {
+	head := s.chainStore.Head()
+	var spState spapow.StoragePowerActorState
+	err := s.chainStore.GetActorStateAt(ctx, head, fcaddr.StoragePowerAddress, &spState)
+	if err != nil {
+		return nil, err
+	}
+
+	infos := []*storagemarket.StorageProviderInfo{}
+	powerHamt, err := hamt.LoadNode(ctx, s.cborStore, spState.PowerTable)
+	if err != nil {
+		return nil, err
+	}
+
+	err = powerHamt.ForEach(ctx, func(minerAddrStr string, _ interface{}) error {
+		fcMinerAddr, err := fcaddr.NewFromString(minerAddrStr)
+		if err != nil {
+			return err
+		}
+
+		var mState spaminer.StorageMinerActorState
+		err = s.chainStore.GetActorStateAt(ctx, head, fcMinerAddr, &mState)
+		if err != nil {
+			return err
+		}
+
+		minerAddr, err := address.NewFromString(minerAddrStr)
+		if err != nil {
+			return err
+		}
+
+		info := mState.Info
+		infos = append(infos, &storagemarket.StorageProviderInfo{
+			Address:    minerAddr,
+			Owner:      info.Owner,
+			Worker:     info.Worker,
+			SectorSize: uint64(info.SectorSize),
+			PeerID:     info.PeerId,
+		})
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return infos, nil
+}
+
+// ValidatePublishedDeal validates a deal has been published correctly
+// Adapted from https://github.com/filecoin-project/lotus/blob/3b34eba6124d16162b712e971f0db2ee108e0f67/markets/storageadapter/client.go#L156
+func (s *StorageClientNodeConnector) ValidatePublishedDeal(ctx context.Context, deal storagemarket.ClientDeal) (uint64, error) {
+	// Fetch receipt to return dealId
+	waitCtx, cancel := context.WithTimeout(ctx, 10*time.Millisecond)
+	defer cancel()
+
+	var publishMsg *types.SignedMessage
+	var publishReceipt *types.MessageReceipt
+
+	err := s.waiter.Wait(waitCtx, *deal.PublishMessage, func(block *block.Block, signedMessage *types.SignedMessage, receipt *types.MessageReceipt) error {
```

Comment:
It feels like you actually want to use `Find` here.  You use 10 milliseconds which tells me this should error if the block with the deal message in it is not already synced by this node, which would make wait the wrong endpoint.

Reply:
(If this was done to avoid annoying cid.Equals(cid) callback you could refactor to `Find` and `FindPredicate`)

Reply:
Good point. I changed it to find.